### ISSUE TITLE
Add a linter to avoid panic in the tree

### ISSUE THREAD

**Issue by joestringer** [cilium/cilium#33364]
In general, we prefer to avoid adding calls to `panic` into production code in the Cilium tree, as crashing the agent at runtime can have significant repercussions on the cluster, including:

- Increasing load to the kube-apiserver due to relisting resources
- Temporary disruption to L7 traffic handled by applications in the agent Pod, including DNS and HTTP (with envoy.enabled=false for in-Pod Envoy)
- Delayed garbage collection for connection tracking tables and DNS history

This issue proposes to add a linter to the tree to avoid adding additional calls to `panic` unless there is a really good reason.

Tasks:
- [ ] Add a linter to golint which prevents new instances of `panic` being introduced into the tree without a nolint comment
- [ ] Review the existing usage of `panic` to outline which instances need to be removed, and which ones can be ignored via a nolint comment
- [ ] Configure the linter to ignore panics in autogenerated api code (which has an exception catcher middleware).
- [ ] Consider providing an alternative function to `panic` which provides log-friendly stack-traces. This should force CI to fail, but it should not end flow control for the function/agent.

**Comment by joestringer** [comment]
From a developer perspective, I think that one of the common motivations for using `panic` is:
- There is a precondition that I expect to never be hit, but if we ever _do_ hit it, then I would like it to be obvious in CI so that I can deal with the issue before it is released.
- By default, `panic` provides a nicely formatted stack-trace that can pinpoint the specific code which causes the issue.

The first item is tempting because if the developer does not need to reason about what happens beyond an error condition (due to `panic`), then it is simpler to reason about and the code can be more quickly written. However, I think that in some such cases we should be taking a step back to consider whether we can structure the code to eliminate the possibility of panic in the first place. For instance, if there is type conversion code for converting a string into an integer somewhere in the middle of the agent, then perhaps we should introduce better sanitisation of the types at the API boundaries so that internally the data is always formatted as an integer.

The latter makes sense, though ideally we should also continue to ensure that logs are properly structured. I've added a task to consider whether we can have a logging helper function that will format the stacktraces properly in log output (for instance using github.com/hashicorp/go-hclog, and individually logging each line).

**Comment by christarazi** [comment]
I agree with the points laid out here. One small thing we might want to the linter to detect as well is `log.Panic` as well.
